File: dart/0155-min-stack.dart
//  Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

void main() {
  var stack = [];
  var minstack = [];

  void push(int value) {
    stack.add(value);
    final minListData = [value, minstack.isEmpty ? value : minstack.last];
    final min = minListData
        .reduce((value, element) => value < element ? value : element);
    minstack.add(min);
  }

  void pop() {
    stack.removeLast();
    minstack.removeLast();
  }

  int getTop() {
    return stack.last;
  }

  int getMin() {
    return minstack.last;
  }

  push(-1);
  push(0);
  push(-3);
  print('Top : ${getTop()}');
  print('Min : ${getMin()}');
  pop();
  print('Top After Pop: ${getTop()}');
  print('Min After Pop: ${getMin()}');
}

File: swift/0155-min-stack.swift
class MinStack {
    var stack: [Int]
    var minStack: [Int]
    init() {
        stack = []
        minStack = []
    }
    
    func push(_ val: Int) {
        stack.append(val)
        if minStack.isEmpty {
            minStack.append(val)
        }
        else {
            let minVal = min(val, minStack.last!)
            minStack.append(minVal)
        }

    }
    
    func pop() {
        stack.removeLast()
        minStack.removeLast()
    }
    
    func top() -> Int {
        return stack.last!
    }
    
    func getMin() -> Int {
        return minStack.last!
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * let obj = MinStack()
 * obj.push(val)
 * obj.pop()
 * let ret_3: Int = obj.top()
 * let ret_4: Int = obj.getMin()
 */